File: i2f-jdk/i2f-swl/web/i2f-swl/src/i2f-turbo-web/i2f-swl/core/core/SwlTransferRefactor.js
import SwlRsaAsymmetricEncryptorSupplier from "../impl/supplier/SwlRsaAsymmetricEncryptorSupplier";
import SwlAesSymmetricEncryptorSupplier from "../impl/supplier/SwlAesSymmetricEncryptorSupplier";
import SwlSha256MessageDigester from "../impl/SwlSha256MessageDigester";
import SwlBase64Obfuscator from "../impl/SwlBase64Obfuscator";
import MemMapExpireCache from "../../../i2f-core/cache/impl/MemMapExpireCache";
import SwlTransferConfig from "./SwlTransferConfig";
import Random from "../../../i2f-core/util/Random";
import AsymKeyPair from "../../../i2f-core/crypto/asymmetric/AsymKeyPair";
import SwlExchanger from "@/i2f-turbo-web/i2f-swl/core/core/exchanger/SwlExchanger";

/**
 * @return SwlTransferRefactor
 * @constructor {SwlTransferRefactor}
 * @extends {SwlExchanger}
 */
function SwlTransferRefactor() {
    SwlExchanger.call(this)
    /**
     *
     * @type {ISwlAsymmetricEncryptorSupplier}
     */
    this.asymmetricEncryptorSupplier = new SwlRsaAsymmetricEncryptorSupplier()
    /**
     *
     * @type {ISwlSymmetricEncryptorSupplier}
     */
    this.symmetricEncryptorSupplier = new SwlAesSymmetricEncryptorSupplier()
    /**
     *
     * @type {ISwlMessageDigester}
     */
    this.messageDigester = new SwlSha256MessageDigester()
    /**
     *
     * @type {ISwlObfuscator}
     */
    this.obfuscator = new SwlBase64Obfuscator()
    /**
     *
     * @type {IExpireCache}
     */
    this.cache = new MemMapExpireCache();
    /**
     *
     * @type {SwlTransferConfig}
     */
    this.config = new SwlTransferConfig()
    /**
     *
     * @type Random
     */
    this.random = new Random()
    /**
     *
     * @type {boolean}
     */
    this.refreshing = false
    /**
     *
     * @type {number}
     */
    this.schedulePool = null
}

// 继承
SwlTransferRefactor.prototype = Object.create(SwlExchanger.prototype)
SwlTransferRefactor.prototype.constructor = SwlTransferRefactor


SwlTransferRefactor.SELF_KEY_PAIR_CURRENT_KEY = function () {
    return "swl:key:self:current"

};
SwlTransferRefactor.SELF_KEY_PAIR_HISTORY_KEY_PREFIX = function () {
    return "swl:key:self:history:"
};
SwlTransferRefactor.OTHER_KEY_PUBLIC_KEY_PREFIX = function () {
    return "swl:key:other:keys:"
};
SwlTransferRefactor.OTHER_KEY_PUBLIC_DEFAULT = function () {
    return "swl:key:other:default"
};
SwlTransferRefactor.KEYPAIR_SEPARATOR = function () {
    return "\n==========\n"
};

/**
 * @return {void}
 */
SwlTransferRefactor.prototype.initialRefreshingThread = function () {
    if (this.refreshing) {
        return
    }
    this.refreshing = true
    let timeout = this.config.selfKeyExpireSeconds - 10;
    if (timeout <= 0) {
        timeout = this.config.selfKeyExpireSeconds;
    }
    let _this = this
    this.schedulePool = setInterval(() => {
        _this.resetSelfKeyPair()
    }, timeout)

    window.onclose = function () {
        if (_this.schedulePool) {
            clearInterval(_this.schedulePool)
            _this.schedulePool = null
        }
    }
}

/**
 *
 * @param key {String}
 * @return {String}
 */
SwlTransferRefactor.prototype.cacheKey = function (key) {
    let cacheKeyPrefix = this.config.cacheKeyPrefix;
    if (cacheKeyPrefix == null || cacheKeyPrefix === "") {
        return key;
    }
    return cacheKeyPrefix + ":" + key;
}

/**
 *
 * @return {AsymKeyPair}
 */
SwlTransferRefactor.prototype.getSelfKeyPair = function () {
    let obj = this.cache.get(this.cacheKey(SwlTransferRefactor.SELF_KEY_PAIR_CURRENT_KEY()));
    if (obj == null) {
        return this.resetSelfKeyPair();
    }
    let key = SwlTransferRefactor.SELF_KEY_PAIR_HISTORY_KEY_PREFIX() + obj;
    obj = this.cache.get(this.cacheKey(key));
    if (obj == null) {
        return this.resetSelfKeyPair();
    }

    let arr = obj.split(SwlTransferRefactor.KEYPAIR_SEPARATOR(), 2);
    let ret = new AsymKeyPair();
    ret.setPublicKey(this.obfuscateDecode(arr[0]));
    ret.setPrivateKey(this.obfuscateDecode(arr[1]));
    return ret;
}

/**
 * @return {String}
 */
SwlTransferRefactor.prototype.getSelfPublicKey = function () {
    let keyPair = this.getSelfKeyPair();
    return keyPair.getPublicKey();
}

/**
 *
 * @param publicKey String
 * @return {String}
 */
SwlTransferRefactor.prototype.calcKeySign = function (publicKey) {
    return this.messageDigester.digest(publicKey);
}

/**
 *
 * @param selfAsymSign {String}
 * @param selfKeyPair {AsymKeyPair}
 * @return {void}
 */
SwlTransferRefactor.prototype.setSelfKeyPair = function (selfAsymSign, selfKeyPair) {
    let key = SwlTransferRefactor.SELF_KEY_PAIR_HISTORY_KEY_PREFIX() + selfAsymSign;
    let keyPair = new AsymKeyPair(
        this.obfuscateEncode(selfKeyPair.getPublicKey()),
        this.obfuscateEncode(selfKeyPair.getPrivateKey())
    );
    let text = keyPair.getPublicKey() + SwlTransferRefactor.KEYPAIR_SEPARATOR() + keyPair.getPrivateKey();
    this.cache.setWith(this.cacheKey(key), text, this.config.selfKeyMaxCount * this.config.selfKeyExpireSeconds);
    this.cache.set(this.cacheKey(SwlTransferRefactor.SELF_KEY_PAIR_CURRENT_KEY()), selfAsymSign);
}

/**
 *
 * @param selfAsymSign {String}
 * @return {String|null}
 */
SwlTransferRefactor.prototype.getSelfPrivateKey = function (selfAsymSign) {
    let key = SwlTransferRefactor.SELF_KEY_PAIR_HISTORY_KEY_PREFIX() + selfAsymSign;
    let obj = this.cache.get(this.cacheKey(key));
    if (obj == null) {
        return null;
    }
    let arr = obj.split(SwlTransferRefactor.KEYPAIR_SEPARATOR(), 2);
    return this.obfuscateDecode(arr[1]);
}

/**
 *
 * @param otherAsymSign {String}
 * @return {String|null}
 */
SwlTransferRefactor.prototype.getOtherPublicKey = function (otherAsymSign) {
    let key = SwlTransferRefactor.OTHER_KEY_PUBLIC_KEY_PREFIX() + otherAsymSign;
    let obj = this.cache.get(this.cacheKey(key));
    if (obj == null) {
        return null;
    }
    return this.obfuscateDecode(obj);
}

/**
 *
 * @return {String|null}
 */
SwlTransferRefactor.prototype.getOtherPublicKeyDefault = function () {
    let obj = this.cache.get(this.cacheKey(SwlTransferRefactor.OTHER_KEY_PUBLIC_DEFAULT()));
    if (obj == null) {
        return null;
    }
    let otherAsymSign = obj;
    return this.getOtherPublicKey(otherAsymSign);
}

/**
 *
 * @param otherAsymSign {String}
 * @param publicKey {String}
 * @return {void}
 */
SwlTransferRefactor.prototype.setOtherPublicKey = function (otherAsymSign, publicKey) {
    let key = SwlTransferRefactor.OTHER_KEY_PUBLIC_KEY_PREFIX() + otherAsymSign;
    this.cache.setWith(this.cacheKey(key), this.obfuscateEncode(publicKey), this.config.otherKeyExpireSeconds);
    this.cache.set(this.cacheKey(SwlTransferRefactor.OTHER_KEY_PUBLIC_DEFAULT()), otherAsymSign);
}

/**
 *
 * @param otherAsymSign {String}
 * @return {void}
 */
SwlTransferRefactor.prototype.refreshOtherPublicKeyExpire = function (otherAsymSign) {
    let key = SwlTransferRefactor.OTHER_KEY_PUBLIC_KEY_PREFIX() + otherAsymSign;
    this.cache.expire(this.cacheKey(key), this.config.otherKeyExpireSeconds);
}


/**
 *
 * @param otherPublicKey {String}
 * @return {void}
 */
SwlTransferRefactor.prototype.acceptOtherPublicKey = function (otherPublicKey) {
    let otherAsymSign = this.messageDigester.digest(otherPublicKey);
    this.setOtherPublicKey(otherAsymSign, otherPublicKey);
}

/**
 *
 * @return {String}
 */
SwlTransferRefactor.prototype.getSelfSwapKey = function () {
    let selfPublicKey = this.getSelfPublicKey();
    return this.obfuscateEncode(selfPublicKey);
}

/**
 *
 * @param otherSwapKey {String}
 * @return {void}
 */
SwlTransferRefactor.prototype.acceptOtherSwapKey = function (otherSwapKey) {
    let otherPublicKey = this.obfuscateDecode(otherSwapKey);
    this.acceptOtherPublicKey(otherPublicKey);
}

/**
 *
 * @return {AsymKeyPair}
 */
SwlTransferRefactor.prototype.resetSelfKeyPair = function () {
    let asymmetricEncryptor = this.asymmetricEncryptorSupplier.get();
    let asymKeyPair = asymmetricEncryptor.generateKeyPair();
    let selfAsymSign = this.messageDigester.digest(asymKeyPair.getPublicKey());
    this.setSelfKeyPair(selfAsymSign, asymKeyPair);
    return asymKeyPair;
}

/**
 *
 * @param remotePublicKey {String}
 * @param parts {string[]}
 * @param attaches {string[]|null}
 * @return {SwlData}
 */
SwlTransferRefactor.prototype.send = function (remotePublicKey, parts, attaches = null) {
    let keyPair = this.getSelfKeyPair();
    return this.sendByRaw(remotePublicKey, this.calcKeySign(remotePublicKey),
        keyPair.getPrivateKey(), this.calcKeySign(keyPair.getPublicKey()),
        parts, attaches);
}

/**
 *
 * @param clientId {String}
 * @param request {SwlData}
 * @return {SwlData}
 */
SwlTransferRefactor.prototype.receive = function (clientId, request) {
    let selfAsymSign = request.header.remoteAsymSign;
    let otherAsymSign = request.header.localAsymSign;
    let otherPublicKey = this.getOtherPublicKey(otherAsymSign);
    let selfPrivateKey = this.getSelfPrivateKey(selfAsymSign);
    return this.receiveByRaw(clientId, request, otherPublicKey, selfPrivateKey);
}

/**
 *
 * @param parts {string[]}
 * @param attaches {string[]|null}
 * @return {SwlData}
 */
SwlTransferRefactor.prototype.sendDefault = function (parts, attaches = null) {
    let otherPublicKey = this.getOtherPublicKeyDefault();
    return this.send(otherPublicKey, parts, attaches);
}

/**
 *
 * @param remoteAsymSign {String}
 * @param parts {string[]}
 * @param attaches {string[]|null}
 * @return {SwlData}
 */
SwlTransferRefactor.prototype.response = function (remoteAsymSign, parts, attaches = null) {
    let otherPublicKey = this.getOtherPublicKey(remoteAsymSign);
    return this.send(otherPublicKey, parts, attaches);
}

export default SwlTransferRefactor

File: i2f-jdk/i2f-swl/src/main/java/i2f/swl/core/SwlTransferRefactor.java
package i2f.swl.core;

import i2f.cache.impl.container.MapCache;
import i2f.cache.impl.expire.ObjectExpireCacheWrapper;
import i2f.cache.std.expire.IExpireCache;
import i2f.crypto.std.encrypt.asymmetric.key.AsymKeyPair;
import i2f.swl.core.exchanger.SwlExchanger;
import i2f.swl.data.SwlData;
import i2f.swl.std.ISwlAsymmetricEncryptor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.List;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * @author Ice2Faith
 * @date 2024/7/9 20:22
 * @desc
 */
@Data

@NoArgsConstructor
public class SwlTransferRefactor extends SwlExchanger {
    public static final String SELF_KEY_PAIR_CURRENT_KEY = "swl:key:self:current";
    public static final String SELF_KEY_PAIR_HISTORY_KEY_PREFIX = "swl:key:self:history:";
    public static final String OTHER_KEY_PUBLIC_KEY_PREFIX = "swl:key:other:keys:";
    public static final String OTHER_KEY_PUBLIC_DEFAULT = "swl:key:other:default";
    public static final String KEYPAIR_SEPARATOR = "\n==========\n";

    private IExpireCache<String, String> cache = new ObjectExpireCacheWrapper<>(new MapCache<>(new ConcurrentHashMap<>()));
    private SwlTransferConfig config = new SwlTransferConfig();

    private AtomicBoolean refreshing = new AtomicBoolean(false);
    private ScheduledExecutorService schedulePool = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
        @Override
        public Thread newThread(Runnable r) {
            Thread thread = new Thread(r, "swl-refreshing");
            thread.setDaemon(true);
            return thread;
        }
    });

    public void initialRefreshingThread() {
        if (refreshing.getAndSet(true)) {
            return;
        }
        long timeout = config.getSelfKeyExpireSeconds() - 10;
        if (timeout <= 0) {
            timeout = config.getSelfKeyExpireSeconds();
        }
        schedulePool.scheduleAtFixedRate(() -> {
            resetSelfKeyPair();
        }, 0, timeout, TimeUnit.SECONDS);
    }

    public String cacheKey(String key) {
        String cacheKeyPrefix = config.getCacheKeyPrefix();
        if (cacheKeyPrefix == null || cacheKeyPrefix.isEmpty()) {
            return key;
        }
        return cacheKeyPrefix + ":" + key;
    }

    public AsymKeyPair getSelfKeyPair() {
        String obj = cache.get(cacheKey(SELF_KEY_PAIR_CURRENT_KEY));
        if (obj == null) {
            return resetSelfKeyPair();
        }
        String key = SELF_KEY_PAIR_HISTORY_KEY_PREFIX + obj;
        obj = cache.get(cacheKey(key));
        if (obj == null) {
            return resetSelfKeyPair();
        }

        String[] arr = obj.split(KEYPAIR_SEPARATOR, 2);
        AsymKeyPair ret = new AsymKeyPair();
        ret.setPublicKey(obfuscateDecode(arr[0]));
        ret.setPrivateKey(obfuscateDecode(arr[1]));
        return ret;
    }

    public String getSelfPublicKey() {
        AsymKeyPair keyPair = getSelfKeyPair();
        return keyPair.getPublicKey();
    }

    public String calcKeySign(String publicKey) {
        return messageDigester.digest(publicKey);
    }

    public void setSelfKeyPair(String selfAsymSign, AsymKeyPair selfKeyPair) {
        String key = SELF_KEY_PAIR_HISTORY_KEY_PREFIX + selfAsymSign;
        AsymKeyPair keyPair = new AsymKeyPair(
                obfuscateEncode(selfKeyPair.getPublicKey()),
                obfuscateEncode(selfKeyPair.getPrivateKey())
        );
        String text = keyPair.getPublicKey() + KEYPAIR_SEPARATOR + keyPair.getPrivateKey();
        cache.set(cacheKey(key), text, config.getSelfKeyMaxCount() * config.getSelfKeyExpireSeconds(), TimeUnit.SECONDS);
        cache.set(cacheKey(SELF_KEY_PAIR_CURRENT_KEY), selfAsymSign);
    }

    public String getSelfPrivateKey(String selfAsymSign) {
        String key = SELF_KEY_PAIR_HISTORY_KEY_PREFIX + selfAsymSign;
        String obj = cache.get(cacheKey(key));
        if (obj == null) {
            return null;
        }
        String[] arr = obj.split(KEYPAIR_SEPARATOR, 2);
        return obfuscateDecode(arr[1]);
    }

    public String getOtherPublicKey(String otherAsymSign) {
        String key = OTHER_KEY_PUBLIC_KEY_PREFIX + otherAsymSign;
        String obj = cache.get(cacheKey(key));
        if (obj == null) {
            return null;
        }
        return obfuscateDecode(obj);
    }

    public String getOtherPublicKeyDefault() {
        String obj = cache.get(cacheKey(OTHER_KEY_PUBLIC_DEFAULT));
        if (obj == null) {
            return null;
        }
        String otherAsymSign = obj;
        return getOtherPublicKey(otherAsymSign);
    }

    public void setOtherPublicKey(String otherAsymSign, String publicKey) {
        String key = OTHER_KEY_PUBLIC_KEY_PREFIX + otherAsymSign;
        cache.set(cacheKey(key), obfuscateEncode(publicKey), config.getOtherKeyExpireSeconds(), TimeUnit.SECONDS);
        cache.set(cacheKey(OTHER_KEY_PUBLIC_DEFAULT), otherAsymSign);
    }

    public void refreshOtherPublicKeyExpire(String otherAsymSign) {
        String key = OTHER_KEY_PUBLIC_KEY_PREFIX + otherAsymSign;
        cache.expire(cacheKey(key), config.getOtherKeyExpireSeconds(), TimeUnit.SECONDS);
    }

    public void acceptOtherPublicKey(String otherPublicKey) {
        String otherAsymSign = messageDigester.digest(otherPublicKey);
        setOtherPublicKey(otherAsymSign, otherPublicKey);
    }

    public String getSelfSwapKey() {
        String selfPublicKey = getSelfPublicKey();
        return obfuscateEncode(selfPublicKey);
    }

    public void acceptOtherSwapKey(String otherSwapKey) {
        String otherPublicKey = obfuscateDecode(otherSwapKey);
        acceptOtherPublicKey(otherPublicKey);
    }

    public AsymKeyPair resetSelfKeyPair() {
        ISwlAsymmetricEncryptor asymmetricEncryptor = asymmetricEncryptorSupplier.get();
        AsymKeyPair asymKeyPair = asymmetricEncryptor.generateKeyPair();
        String selfAsymSign = messageDigester.digest(asymKeyPair.getPublicKey());
        setSelfKeyPair(selfAsymSign, asymKeyPair);
        return asymKeyPair;
    }

    public SwlData send(String remotePublicKey, List<String> parts) {
        return send(remotePublicKey, parts, null);
    }

    public SwlData send(String remotePublicKey, List<String> parts, List<String> attaches) {
        AsymKeyPair keyPair = getSelfKeyPair();
        return sendByRaw(remotePublicKey, calcKeySign(remotePublicKey),
                keyPair.getPrivateKey(), calcKeySign(keyPair.getPublicKey()),
                parts, attaches);
    }

    public SwlData receive(String clientId, SwlData request) {
        String selfAsymSign = request.getHeader().getRemoteAsymSign();
        String otherAsymSign = request.getHeader().getLocalAsymSign();
        String otherPublicKey = getOtherPublicKey(otherAsymSign);
        String selfPrivateKey = getSelfPrivateKey(selfAsymSign);
        return receiveByRaw(clientId, request, otherPublicKey, selfPrivateKey);
    }

    public SwlData sendDefault(List<String> parts, List<String> attaches) {
        String otherPublicKey = getOtherPublicKeyDefault();
        return send(otherPublicKey, parts, attaches);
    }

    public SwlData sendDefault(List<String> parts) {
        String otherPublicKey = getOtherPublicKeyDefault();
        return send(otherPublicKey, parts, null);
    }

    public SwlData response(String remoteAsymSign, List<String> parts, List<String> attaches) {
        String otherPublicKey = getOtherPublicKey(remoteAsymSign);
        return send(otherPublicKey, parts, attaches);
    }

    public SwlData response(String remoteAsymSign, List<String> parts) {
        String otherPublicKey = getOtherPublicKey(remoteAsymSign);
        return send(otherPublicKey, parts, null);
    }
}
